[SYCL] Add support for -foffload-fp32-prec-div/sqrt options.

zahiraam opened the pull request:
Add support for options `-f[no]-offload-fp32-prec-div` and `-f[no-]-offload-fp32-prec-sqrt`.
These options are added to allow users to control whether `fdiv` and `sqrt` operations in offload device code are required to return correctly rounded results. In order to communicate this to the device code, we need the front end to generate IR that reflects the choice.

When the correctly rounded setting is used, we can just generate the `fdiv` instruction and `llvm.sqrt` intrinsic, because these operations are required to be correctly rounded by default in LLVM IR.

When the result is not required to be correctly rounded, the front end should generate a call to the `llvm.fpbuiltin.fdiv` or `llvm.fpbuiltin.sqrt` intrinsic with the `fpbuiltin-max-error` attribute set. For single precision` fdiv`, the setting should be `2.5`. For single-precision sqrt, the setting should be `3.0`.

If the -ffp-accuracy option is used, we should issue warnings if the settings conflict with an explicitly set `-foffload-fp32-prec-div` or `-foffload-fp32-prec-sqrt` option.

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -1782,6 +1818,10 @@ void Clang::RenderTargetOptions(const llvm::Triple &EffectiveTriple,
   default:
     break;
 
+  case llvm::Triple::spir:
+  case llvm::Triple::spir64:
```
Please add `llvm::Triple::spirv32` and `llvm::Triple::spirv64`

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -1721,6 +1721,40 @@ static void CollectARMPACBTIOptions(const ToolChain &TC, const ArgList &Args,
+                              const JobAction &JA, const Driver &D) const {
+  if (JA.isDeviceOffloading(Action::OFK_SYCL)) {
+    if (Arg *A = Args.getLastArg(options::OPT_ffp_model_EQ)) {
+      if (!strcmp(A->getValue(), "fast")) {
+        CmdArgs.push_back("-fno-offload-fp32-prec-div");
+        CmdArgs.push_back("-fno-offload-fp32-prec-sqrt");
+      }
```
Should we allow users to override with `-foffload-fp32-prec-div|sqrt`?
```suggestion
      if (!strcmp(A->getValue(), "fast")) {
        if (!Args.hasFlag(option::OPT_foffload_fp32_prec_div,
                          option::OPT_fno_offload_fp32_prec_div, false))
          CmdArgs.push_back("-fno-offload-fp32-prec-div");
        if (!Args.hasFlag(option::OPT_foffload_fp32_prec_sqrt,
                          option::OPT_fno_offload_fp32_prec_sqrt, false))
          CmdArgs.push_back("-fno-offload-fp32-prec-sqrt");
      }
```

zahiraam replied:
Not sure. I would think that users could choose to compile with:
`clang -fsycl -ffp-model=fast -foffload-fp32-prec-sqrt hello.cpp`
or:
`clang -fsycl -foffload-fp32-prec-sqrt -ffp-model=fast hello.cpp`
These shouldn't give the same result. In the first one, the sqrt results are precise. In the second one, they are rounded.

I think that's just following the last command wins rule. In which case we need a compilated process here to find the order in which the options interact with one another.

mdtoguchi replied:
Hmm... If that's the case we may want to integrate the logic into where all of the other FP model options are being manipulated (the larger for loop here: https://github.com/intel/llvm/blob/bdf78d7d165f9022397f885590c4889eb6513682/clang/lib/Driver/ToolChains/Clang.cpp#L2994) and only add the `-cc1` option under the `IsDeviceOffloading` condition.

zahiraam replied:
Okay and that would work for OpenMP too!

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -1721,6 +1721,40 @@ static void CollectARMPACBTIOptions(const ToolChain &TC, const ArgList &Args,
+      if (Args.getLastArg(options::OPT_fno_offload_fp32_prec_sqrt))
+        EmitAccuracyDiag(D, JA, A->getValue(), "-fno-offload-fp32-prec-sqrt");
+    }
+    if (Args.getLastArg(options::OPT_fno_offload_fp32_prec_div))
+      CmdArgs.push_back("-fno-offload-fp32-prec-div");
+    else
+      CmdArgs.push_back("-foffload-fp32-prec-div");
```
```suggestion
    if (!Args.hasFlag(option::OPT_foffload_fp32_prec_div,
                      option::OPT_fno_offload_fp32_prec_div, true))
      CmdArgs.push_back("-fno-offload-fp32-prec-div");
```
Since `-foffload-fp32-prec-div` is default

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -1721,6 +1721,40 @@ static void CollectARMPACBTIOptions(const ToolChain &TC, const ArgList &Args,
+    if (Args.getLastArg(options::OPT_fno_offload_fp32_prec_sqrt))
+      CmdArgs.push_back("-fno-offload-fp32-prec-sqrt");
+    else
+      CmdArgs.push_back("-foffload-fp32-prec-sqrt");
```
similar comment to above.

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3027,6 +3037,13 @@ static void RenderFloatingPointOptions(const ToolChain &TC, const Driver &D,
     Range = LangOptions::ComplexRangeKind::CX_Basic;
     SeenUnsafeMathModeOption = true;
+    if (JA.isDeviceOffloading(Action::OFK_SYCL) ||
+        JA.isDeviceOffloading(Action::OFK_OpenMP)) {
```
Can we use a boolean at the top of the function that covers?
`bool IsDeviceOffloading = JA.isDeviceOffloading(Action::OFK_SYCL);`
When we pulldown these to xmain, we can add the `OFK_OpenMP` check as part of this.

mdtoguchi commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1157,6 +1157,22 @@ defm cx_fortran_rules: BoolOptionWithoutMarshalling<"f", "cx-fortran-rules",
   NegFlag<SetFalse, [], [ClangOption, CC1Option], "Range reduction is disabled "
   "for complex arithmetic operations">>;
 
+ defm offload_fp32_prec_div: BoolOption<"f", "offload-fp32-prec-div",
+  LangOpts<"OffloadFp32PrecDiv">, DefaultTrue,
+  PosFlag<SetTrue, [], [ClangOption, CC1Option], "fdiv operations in offload device "
+  "code are required to return correctly rounded results.">,
+  NegFlag<SetFalse, [], [ClangOption, CC1Option], "fdiv operations in offload device "
+  "code are not required to return correctly rounded results.">>,
+  Group<f_Group>;
+
+ defm offload_fp32_prec_sqrt: BoolOption<"f", "offload-fp32-prec-sqrt",
+ LangOpts<"OffloadFp32PrecSqrt">, DefaultTrue,
+  PosFlag<SetTrue, [], [ClangOption, CC1Option], "sqrt operations in offload device "
+  "code are required to return correctly rounded results.">,
+  NegFlag<SetFalse, [], [ClangOption, CC1Option], "sqrt operations in offload device "
+  "code are not required to return correctly rounded results.">>,
+  Group<f_Group>;
```
```suggestion
 defm offload_fp32_prec_div: BoolOption<"f", "offload-fp32-prec-div",
   LangOpts<"OffloadFp32PrecDiv">, DefaultTrue,
   PosFlag<SetTrue, [], [ClangOption, CC1Option], "fdiv operations in offload device "
   "code are required to return correctly rounded results.">,
   NegFlag<SetFalse, [], [ClangOption, CC1Option], "fdiv operations in offload device "
   "code are not required to return correctly rounded results.">>,
   Group<f_Group>;

 defm offload_fp32_prec_sqrt: BoolOption<"f", "offload-fp32-prec-sqrt",
 LangOpts<"OffloadFp32PrecSqrt">, DefaultTrue,
   PosFlag<SetTrue, [], [ClangOption, CC1Option], "sqrt operations in offload device "
   "code are required to return correctly rounded results.">,
   NegFlag<SetFalse, [], [ClangOption, CC1Option], "sqrt operations in offload device "
   "code are not required to return correctly rounded results.">>,
   Group<f_Group>;
```

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -1781,7 +1781,6 @@ void Clang::RenderTargetOptions(const llvm::Triple &EffectiveTriple,
   switch (TC.getArch()) {
   default:
     break;
-
```
Inadvertent line removal?

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.h:
```diff
@@ -55,6 +55,9 @@ class LLVM_LIBRARY_VISIBILITY Clang : public Tool {
                         bool KernelOrKext) const;
+  void AddSPIRTargetArgs(const llvm::opt::ArgList &Args,
+                         llvm::opt::ArgStringList &CmdArgs, const JobAction &JA,
+                         const Driver &D) const;
```
Changes not needed anymore?

MrSidims reviewed (commented):
Not sure if it should go to this PR (guess it should). The feature uses SPV_INTEL_fp_max_error SPIR-V extension. Currently it's enabled only for AOT CPU compilation. So now we should also enable it when the options are passed.

zahiraam commented:
@MrSidims Are you saying these new options should only be enabled with `SPV_INTEL_fp_max_error` (I am not familiar with this option). I see that this latter option is used only for CPU compilation https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/ToolChains/Clang.cpp#L10742.
The options implemented in this patch are for device compilation only.

zahiraam commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1889,15 +1889,23 @@ void CodeGenModule::getDefaultFunctionFPAccuracyAttributes(
+    if (Name == "fdiv" && !getLangOpts().OffloadFp32PrecDiv)
+      FPAccuracyVal = "2.5";
+    if (!FPAccuracyVal.empty())
+      FuncAttrs.addAttribute("fpbuiltin-max-error", FPAccuracyVal);
```
If the condition in 1894 is satisfied, then the `FuncAttrs.size() != 0)`; we will not get into this code.

zahiraam replied:
Oh! good catch. I think this will fix it.

zahiraam commented:
Thanks @mdtoguchi for the explanation. I will add the extension to this PR.

MrSidims commented:
@zahiraam I believe https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/ToolChains/Clang.cpp#L10742 should be changed to something like:
`if (IsCPU || non-precise-div-opt-enabled || non-precise-sqrt-opt-enabled)`

mdtoguchi commented on clang/test/Driver/sycl-spirv-ext-old-model.c:
```diff
@@ -24,6 +24,63 @@
+// RUN: %clang -target x86_64-unknown-linux-gnu -fsycl --no-offload-new-driver -fsycl-targets=spir64_x86_64-unknown-unknown -fno-offload-fp32-prec-div -fno-offload-fp32-prec-sqrt %s -### 2>&1 \
+// RUN:  | FileCheck %s -check-prefixes=CHECK-CPU
+// RUN: %clang -target x86_64-unknown-linux-gnu -fsycl --no-offload-new-driver -fsycl-targets=spir64_x86_64-unknown-unknown -foffload-fp32-prec-sqrt %s -### 2>&1 \
+// RUN:  | FileCheck %s -check-prefixes=CHECK-CPU-NFPME
```
To reduce the duplication of all of the extensions, can you do something like:  `-check-prefixes=CHECK-CPU,CHECK-CPU-NFPME`
Here, `CHECK-CPU` does not add the fp_max_error, but is rather checked with a `CHECK-CPU-NFPME-NOT`

mdtoguchi commented on clang/test/Driver/sycl-spirv-ext-old-model.c:
```diff
@@ -24,6 +24,63 @@
+// RUN: %clang -target x86_64-unknown-linux-gnu -fsycl --no-offload-new-driver -fsycl-targets=spir64_x86_64-unknown-unknown -fno-offload-fp32-prec-sqrt -fno-offload-fp32-prec-div %s -### 2>&1 \
+// RUN:  | FileCheck %s -check-prefixes=CHECK-CPU
+// RUN: %clang -target x86_64-unknown-linux-gnu -fsycl --no-offload-new-driver -fsycl-targets=spir64_x86_64-unknown-unknown -fno-offload-fp32-prec-div -fno-offload-fp32-prec-sqrt %s -### 2>&1 \
+// RUN:  | FileCheck %s -check-prefixes=CHECK-CPU
```
To reduce the duplication of all of the extensions, can you do something like: `-check-prefixes=CHECK-CPU,CHECK-CPU-FPME`
Here, `CHECK-CPU` does not add the fp_max_error, but is rather checked with a `CHECK-CPU-FPME`

mdtoguchi commented on clang/test/Driver/sycl-spirv-ext-old-model.c:
```diff
@@ -129,3 +186,110 @@
 // CHECK-CPU-SAME:,+SPV_KHR_non_semantic_info
 // CHECK-CPU-SAME:,+SPV_KHR_cooperative_matrix
 // CHECK-CPU-SAME:,+SPV_INTEL_fp_max_error"
```
```suggestion
// CHECK-CPU-FPME:,+SPV_INTEL_fp_max_error"
```
To match the suggested change with `-check-prefixes` above.

mdtoguchi commented on clang/test/Driver/sycl-spirv-ext-old-model.c:
```diff
@@ -129,3 +186,110 @@
 // CHECK-CPU-SAME:,+SPV_KHR_cooperative_matrix
 // CHECK-CPU-SAME:,+SPV_INTEL_fp_max_error"
+//
+// CHECK-CPU-NFPME: llvm-spirv{{.*}}"-spirv-allow-unknown-intrinsics=llvm.genx.,llvm.fpbuiltin"
+// CHECK-CPU-NFPME-SAME: {{.*}}"-spirv-ext=-all
+// CHECK-CPU-NFPME-SAME:,+SPV_EXT_shader_atomic_float_add
+// CHECK-CPU-NFPME-SAME:,+SPV_EXT_shader_atomic_float_min_max
+// CHECK-CPU-NFPME-SAME:,+SPV_KHR_no_integer_wrap_decoration,+SPV_KHR_float_controls
+// CHECK-CPU-NFPME-SAME:,+SPV_KHR_expect_assume,+SPV_KHR_linkonce_odr
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_subgroups,+SPV_INTEL_media_block_io
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_device_side_avc_motion_estimation
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_fpga_loop_controls
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_unstructured_loop_controls,+SPV_INTEL_fpga_reg
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_blocking_pipes,+SPV_INTEL_function_pointers
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_kernel_attributes,+SPV_INTEL_io_pipes
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_inline_assembly,+SPV_INTEL_arbitrary_precision_integers
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_float_controls2
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_vector_compute,+SPV_INTEL_fast_composite
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_arbitrary_precision_fixed_point
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_arbitrary_precision_floating_point
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_variable_length_array,+SPV_INTEL_fp_fast_math_mode
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_long_constant_composite
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_arithmetic_fence
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_cache_controls
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_fpga_buffer_location
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_fpga_argument_interfaces
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_fpga_invocation_pipelining_attributes
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_fpga_latency_control
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_task_sequence
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_token_type
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_bfloat16_conversion
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_joint_matrix
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_hw_thread_queries
+// CHECK-CPU-NFPME-SAME:,+SPV_KHR_uniform_group_instructions
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_masked_gather_scatter
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_tensor_float32_conversion
+// CHECK-CPU-NFPME-SAME:,+SPV_INTEL_optnone
+// CHECK-CPU-NFPME-SAME:,+SPV_KHR_non_semantic_info
+// CHECK-CPU-NFPME-SAME:,+SPV_KHR_cooperative_matrix
+// CHECK-CPU-NFPME-NOT:,+SPV_INTEL_fp_max_error"
```
```suggestion
// CHECK-CPU-NFPME-NOT:,+SPV_INTEL_fp_max_error"
```
To match up with the `-check-prefixes` suggestion above.

mdtoguchi commented on clang/test/Driver/sycl-spirv-ext-old-model.c:
```diff
@@ -129,3 +186,110 @@
+// CHECK-FPGA-HW-FPME-SAME:,+SPV_INTEL_fpga_dsp_control
+// CHECK-FPGA-HW-FPME-SAME:,+SPV_INTEL_fpga_memory_accesses
+// CHECK-FPGA-HW-FPME-SAME:,+SPV_INTEL_fpga_memory_attributes
+// CHECK-FPGA-HW-FPME-SAME:,+SPV_INTEL_fp_max_error"
```
Similar changes here as suggested above to reduce string redundancy.

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -10603,8 +10681,22 @@ static void getTripleBasedSPIRVTransOpts(Compilation &C,
+            TCArgs.hasFlag(options::OPT_fno_offload_fp32_prec_div,
+                           options::OPT_foffload_fp32_prec_div, false));
+  };
+  if (IsCPU && hasNoOffloadFP32PrecOption(TCArgs) ||
+      !IsCPU && shouldUseOffloadFP32PrecOption(TCArgs)) {
```
```suggestion
  if ((IsCPU && hasNoOffloadFP32PrecOption(TCArgs)) ||
      shouldUseOffloadFP32PrecOption(TCArgs)) {
```
I believe the option settings should always trigger regardless if doing AOT for CPU.

premanandrao commented on clang/test/CodeGenSYCL/offload-fp32-div-sqrt.cpp:
```diff
@@ -0,0 +1,229 @@
+// DEFINE: -fsycl-is-device -emit-llvm -triple spirv-unknown-unknown
+
+// DEFINE: %{common_opts_spir64} = -internal-isystem %S/Inputs \
+// DEFINE: -fsycl-is-device -emit-llvm -triple spirv64-unknown-unknown
```
``common_opts_spir64`` seems identical to ``common_opts_spirv64``.

zahiraam replied:
Fixed now.

MrSidims reviewed (approved):
SPV_INTEL_fp_max_error related changes LGTM

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -3055,11 +3073,57 @@ static void RenderFloatingPointOptions(const ToolChain &TC, const Driver &D,
+    }
+  };
+
+  auto ParseFPAccOption = [&](StringRef Val, bool &NoOffloadFlag) {
```
```suggestion
  auto parseFPAccOption = [&](StringRef Val, bool &NoOffloadFlag) {
```
Function naming should start with lowercase.

MrSidims reviewed (changes requested):
Has to withdraw my review as have 2 questions.

MrSidims commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -5790,10 +5809,20 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
+          (FuncName == "sqrt" && !getLangOpts().OffloadFP32PrecSqrt &&
+           IsFloat32Type);
+      bool isFP32FdivFunction =
+          (FuncName == "fdiv" && !getLangOpts().OffloadFP32PrecDiv &&
```
@gmlueck could you please comment on that?

gmlueck replied:
The intent of `-foffload-fp32-prev-div` is to affect the native divide operation (i.e. `/`).  There is no SYCL function named `fdiv`.  Is there a standard C / C++ function with that name?

MrSidims replied:
AFAIK there is no standard function float FP division. There is std::div, but it works only on integers.

zahiraam replied:
There is no C/C++ fdiv function.

zahiraam commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -5790,10 +5809,20 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
+      bool hasFPAccuracyFuncMap = hasAccuracyRequirement(FuncName);
+      bool hasFPAccuracyVal = !getLangOpts().FPAccuracyVal.empty();
+      bool isFp32SqrtFunction =
+          (FuncName == "sqrt" && !getLangOpts().OffloadFP32PrecSqrt &&
```
FuncName is the output of FD->getName() which returns a simple identifier. https://github.com/intel/llvm/blob/sycl/clang/include/clang/AST/Decl.h#L280

MrSidims replied:
So clang/test/CodeGenSYCL/offload-fp32-div-sqrt.cpp with pass even with `extern "C"` removed from `sqrt` function declaration?

gmlueck replied:
What if the user has a function in their own namespace that happens to be named "sqrt"?

mdtoguchi commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -2998,6 +3020,9 @@ static void RenderFloatingPointOptions(const ToolChain &TC, const Driver &D,
   LangOptions::ComplexRangeKind Range = LangOptions::ComplexRangeKind::CX_None;
   std::string ComplexRangeStr = "";
   std::string GccRangeComplexOption = "";
+  bool IsFp32PrecDivSqrtAllowed = JA.isDeviceOffloading(Action::OFK_SYCL) &&
+                                  !JA.isDeviceOffloading(Action::OFK_Cuda) &&
+                                  !JA.isOffloading(Action::OFK_HIP);
```
As discussed offlne, something like:
```suggestion
  bool IsFp32PrecDivSqrtAllowed = JA.isDeviceOffloading(Action::OFK_SYCL) &&
                                  TC.getTriple().isSPIROrSPIRV();
```

zahiraam commented:
@MrSidims and @gmlueck I have removed the restriction for `CUDA/HIP`. My understanding is that @MrSidims will make changes so that the precision for `3.0` is allowed for the `sqrt` function. Is that the case?
I have also changed the code so that `/` has precision set with the options instead of `fdiv`. 
Please let me know if these are the changes you expected.

MrSidims commented on clang/test/CodeGenSYCL/offload-fp32-div-sqrt.cpp:
```diff
@@ -0,0 +1,354 @@
+      // ROUNDED-SQRT-PREC-DIV: call reassoc nnan ninf nsz arcp afn float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT:[0-9]+]]
+      // ROUNDED-DIV-PREC-SQRT: call reassoc nnan ninf nsz arcp afn spir_func nofpclass(nan inf) float @sqrt(float noundef nofpclass(nan inf) {{.*}})
+      // ROUNDED-DIV-ROUNDED-SQRT-FAST: call reassoc nnan ninf nsz arcp afn float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT:[0-9]+]]
+      // LOW-PREC-DIV: call float @llvm.fpbuiltin.sqrt.f32(float {{.*}}) #[[ATTR_SQRT_LOW:[0-9]+]]
```
Just want to check if I understand this correctly. In this case we pass: -fno-offload-fp32-prec-div -ffp-builtin-accuracy=high flags. And 1.0 ULP fpbuiltin-error attribute for llvm.fpbuiltin.sqrt.f32 was generated in response of ffp-builtin-accuracy=high flag and we don't expect here precise calculations as high != precise, right?

zahiraam replied:
That's correct.

MrSidims commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1879,25 +1879,44 @@ void CodeGenModule::getDefaultFunctionFPAccuracyAttributes(
-      StringRef FPAccuracyVal = llvm::fp::getAccuracyForFPBuiltin(
-          ID, FuncType, convertFPAccuracy(getLangOpts().FPAccuracyVal));
+      StringRef FPAccuracyVal;
+      if (!getLangOpts().OffloadFP32PrecDiv && Name == "div")
```
I'm not that familiar with the code, so would like to double check - "div" is just an alias to "llvm.fpbuiltin.fdiv" or is it some function call `@div(...)`. If it's an intrinsic, then why renaming from `fdiv` to `div` was required?

Basically I'm asking if these lines may affect C stdlib's `div` function that have 'long integer' operands. If you say that this code won't affect it - that would be enough explanation for me :)

zahiraam replied:
Yes `div` is just an alias. You can look at the code in `ScalarExprEmitter::EmitDiv` at about line #3788 that the intrinsic's ID is set to `llvm::Intrinsic::fpbuiltin_fdiv`. The call to `CreateBuiltinCallWithAttr` is made with the hard code name `div`.
In order to remove this confusion. I will name it back to `fdiv` as it was. I thought that it would be clearer to rename it `div` so that we know that this is the `/` function instead of the `fdiv` function that responds to these options.

MrSidims reviewed (commented):
LGTM assuming that the code doesn't affect C stdlib's `div` function, see the comment above.

zahiraam commented:
The FE work has been approved. But this is an attempt to fix the LIT test `DeviceLib/cmath_test.cpp`. I am not really sure the change is correct. Will revert it LIT fail persists.

zahiraam commented:
@MrSidims Your PR fixed the issue with `cmath_test.cp`. Thanks!
@intel/llvm-gatekeepers Can this be merged in please? Thanks.

MrSidims commented on clang/test/Driver/offload-fp32-div-sqrt.cu:
```diff
@@ -0,0 +1,39 @@
+
+// CHECK-NOT: "-foffload-fp32-prec-div"
+// CHECK-NOT: "-foffload-fp32-prec-sqrt"
+// FPACC: "-ffp-builtin-accuracy=high"
```
@zahiraam @mdtoguchi hi, I'm revisiting some parts of the feature implementation, to ensure, that everything works right. Here I have a question: if I'm not mistaken -ffp-builtin-accuracy=high means, that all fp-builtins will be calculated with high precision. Shouldn't we restrict it to only sqrt/div fpbuiltins (see example in clang/test/Driver/offload-fp32-div-sqrt.cpp from above)?

@npmiller I see that long ago you have added a similar option "fsycl-fp32-prec-sqrt" that currently works for CUDA and HIP. I do believe the appropriate option added in this patch should be merged with your option (if you agree or don't agree with it please place your thoughts in https://github.com/intel/llvm/pull/17033 ), but I've also noted, that way of handling those options are very different. In your patch you pass to the compiler an alias to -prec-sqrt=true to compiler resulting in metadata (that, I guess), is handled by CUDA backend. This patch instead makes clang to generate `llvm.fpbuiltin.sqrt` with max-error=0.5 to be generated in the response of this option, which is being lowered to `llvm.sqrt` by code added [here](https://github.com/intel/llvm/pull/16714 , which seem to be also a correct approach: https://godbolt.org/z/5TKxbzcKE . If we decide to merge the options, guess we would either need to remove one of the mechanisms for CUDA or duplicate them, what would be your preference?

zahiraam replied:
@MrSidims The options `-f[no]offload-fp32-prec-div/sqrt` are only available when using `-fsycl`. In a cuda environment the options have no effect.

npmiller replied:
I think it would make sense to merge the options.

However, we need to keep the existing way the flag is implemented, because CUDA ships with a bitcode library that we link against our kernels, which provides implementations for some of the math built-ins.

That library makes use of the metadata setup by the current way the flag is handled to pick out between precise and approx versions of sqrt in *different* math built-ins. So we could maybe use `llvm.fpbuiltin.sqrt` for direct calls to `sqrt` from the kernel, but we'll still need to setup the metadata for the bitcode library, so it's probably easier to keep it as-is.

Note that there's a similar mechanism for HIP, although I'm not sure we use it as much, but we probably should keep the current setup there to make sure the metadata is set properly.

MrSidims replied:
@npmiller thanks!

@zahiraam may be I'm not appropriately putting '=' between CUDA and NVPTX environments for SYCL or missing something else. But we had observed a regression in one of CUDA math tests with this very patch, hence the options have affect on it, aren't they (at least -fno-offload-fp32-prec)?

zahiraam replied:
@MrSidims If the expected behavior for the failing test is to generate `llvm-fpbuiltin.sqrt.f32` and `llvm-fpbuiltin.fdiv.f32`, then we should have `fno-offload-fp32-prec-div/sqrt` expand to `cc1` option `fno-offload-fp32-prec-div/sqrt` when compiling `cuda` code. Currently that is the case only when compiling `sycl` code.

MrSidims replied:
@zahiraam I mean that in this very PR in [this pre-commit](https://github.com/intel/llvm/actions/runs/11744956047/job/32722703231) run the following test: SYCL :: sycl-in-tree/DeviceLib/~cmath_test.cpp in CUDA environment has failed due to llvm.fpbuiltin.fdiv.f32 not having lowering in AltMathLib. Thus I make a conclusion, that these options like they are defined now has affect on CUDA, aren't they? What I was trying to understand if such affect is limited only to intrinsics generation or "// FPACC: "-ffp-builtin-accuracy=high"" (from this test) is also part of this effect.

MrSidims replied:
Discussed offline. f[no]-offload-fp32-prec-div/sqrt is not responsible for any further options propagation for CUDA and HIP. It only affects whether llvm.fpbuiltin.div/sqrt with max-error=2.5/3.0 is generated or 'standard' LLVM instructions/intrinsics/builtins will be used.
